### PR TITLE
Fix a buffer overrun in GC structures

### PR DESCRIPTION
**Pull request by AntonLapounov**
Fix two issues in the `record_changed_seg` function:
* It would skip every 129th record.
* In debug builds the value of the `saved_changed_segs_count` global variable is refetched from the memory, and the `saved_changed_segs_count++;` statement makes it possible for the value to exceed 127, resulting in a corruption of GC data structures following the `saved_changed_segs` array in the memory.

The `saved_changed_segs` array was probably introduced for debugging purposes; the saved records are never used by the GC code.  Let me know if I should put this code under `#ifdef _DEBUG`.

### CODE REVIEW

**Comment by Maoni0** [comment]
good find. this used to never need an interlock for segments but for regions we do need the interlock. thanks for the fix!

this is actually used mostly for retail. we use it for diagnostics purposes.

File: src/coreclr/gc/gccommon.cpp [hunk 1]
```diff
@@ -91,26 +91,27 @@ struct changed_seg
...
-        saved_changed_segs[saved_changed_segs_count].gc_index = current_gc_index;
-        saved_changed_segs[saved_changed_segs_count].bgc = current_bgc_state;
-        saved_changed_segs[saved_changed_segs_count].changed = changed_state;
-        saved_changed_segs_count++;
-    }
-    else
-    {
-        saved_changed_segs_count = 0;
-    }
+#ifdef MULTIPLE_HEAPS
```
Comment (Maoni0) [inline src/coreclr/gc/gccommon.cpp:101]: I would make this `#if defined(MULTIPLE_HEAPS) && defined(USE_REGIONS)` and make `saved_changed_segs_count` a 64-bit integer.

File: src/coreclr/gc/gccommon.cpp [hunk 2]
```diff
@@ -91,26 +91,27 @@ struct changed_seg
...
+    int segs_index = Interlocked::Increment(&saved_changed_segs_count);
```
Comment (VSadov) [inline src/coreclr/gc/gccommon.cpp:102]: What happens when 128+ heaps?

Comment (AntonLapounov) [reply src/coreclr/gc/gccommon.cpp:102]: That is a circular buffer, so the oldest entry will be overwritten.

**Review by Maoni0 — APPROVED** [review]

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!!

**Pull request merged**